### ISSUE TITLE
Target Dropout not scaled by keep_prob

### ISSUE THREAD

**Issue by guglielmocamporese** [for-ai/TD#14]
https://github.com/for-ai/TD/blob/877e3b9d1491e78e4e7f1d2e3622983540e415e7/models/utils/dropouts.py#L236

Is there a reason why here you do not scale the input (divide by the probability of keeping a neuron) in order to have the same energy on the layer during training and testing? The standard dropout rescales the input by keep_prob: 

`tf.nn.dropout(...)` -> https://www.tensorflow.org/versions/r1.15/api_docs/python/tf/nn/dropout

Thank you!

**Comment by aidangomez** [comment]
We do this because with normal dropout the intention is to replace the deleted neurons back into the network -- which changes the expected norm of the activation -- so we need to scale the neurons so that when the deleted ones are replaced the expected norm stays about the same.

In our case, we don't intend to return the deleted weights/neurons so there's no need to perform and rescaling.

**Comment by guglielmocamporese** [comment]
Thank you! 

What I noticed is that targeted dropout does not preserve the same expected value of weight/neuron at train and test time:

- during training you sort under a certain rule the neurons/weights and you perform dropout on the "queue",

- during test you sort under the same rule the neurons/weights but you drop deterministically the "queue",

so the average values for neurons/weights is not the same on train and test.

**Comment by aidangomez** [comment]
That's a good point, there will be some shift in this case and TD will likely stand to benefit from performing some sort of rescaling of the neurons that fall within the "queue".

Would be nice to see some experiments comparing the effect of this.

**Comment by guglielmocamporese** [comment]
By the way I just applied at FOR.ai!

**Comment by aidangomez** [comment]
That's great to hear! Looking forward to reviewing your application.
